### PR TITLE
Audit for immediate effect of settings

### PR DESCRIPTION
**Pull request by andreeis**
Ensure immediate effect for settings: cmakeCommunicationMode, generator and preferredGenerators.
More to be investigated and eventually fixed in a following commit.

Also reverted #1417 since we need to fix it in a different way.
Opened #1498 to track the alternative fix of the above PR.

### CODE REVIEW

File: src/cmake-tools.ts [hunk 1]
```diff
@@ -152,6 +152,21 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
   private readonly _codeModel = new Property<codemodel_api.CodeModelContent|null>(null);
   private _codeModelDriverSub: vscode.Disposable|null = null;
 
+  private readonly _communicationModeSub = this.workspaceContext.config.onChange('cmakeCommunicationMode', () => {
```
Comment (bobbrow) [inline src/cmake-tools.ts:155]: I don't really know what 'sub' means here.  I know it's reused from other places, but it doesn't really make sense to me.  Can you open an issue to rename some of these things?

File: src/cmake-tools.ts [hunk 2]
```diff
@@ -152,6 +152,21 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
...
+  private readonly _generatorSub = this.workspaceContext.config.onChange('generator', () => {
+    log.info(localize('generator.changed.restart.driver', "Restarting the CMake driver after a generator change."));
+    return this._reloadCMakeDriver();
+  });
+
+  private readonly _preferredGeneratorsSub = this.workspaceContext.config.onChange('preferredGenerators', () => {
+    log.info(localize('preferredGenerator.changed.restart.driver', "Restarting the CMake driver after a preferredGenerators change."));
+    return this._reloadCMakeDriver();
+  });
+
```
Comment (bobbrow) [inline src/cmake-tools.ts:169]: I expected to see watchers for configureArgs/Settings.  I noticed there is something in driver.ts that should be doing it, but there are still reports of those config settings not initiating a reload of the driver.  Are you able to repro that at all?

**Review by bobbrow — CHANGES_REQUESTED** [review]

File: src/cmake-tools.ts (hunk 2 above)
Comment (andreeis) [reply src/cmake-tools.ts:169]: No, I saw the reports and I tried but I wasn't able to reproduce. They seemed to update fine for me. I tried windows and linux, fileApi and serverApi. Can you also try to reproduce configureArgs/Settings not getting updated between a settings change and a configure?

File: src/cmake-tools.ts (hunk 1 above)
Comment (andreeis) [reply src/cmake-tools.ts:155]: https://github.com/microsoft/vscode-cmake-tools/issues/1509

File: src/cmake-tools.ts (hunk 2 above)
Comment (bobbrow) [reply src/cmake-tools.ts:169]: It looks like there may be a bug with deleting the whole setting not taking effect, but I'm not going to block this PR on that.  We can open a bug for it.  The important scenario of being able to add settings to a fresh folder appears to be working, so I'll approve.

**Pull request merged**